cpp/src/arrow/scalar.cc

```diff
@@ -310,6 +316,19 @@ struct MakeNullImpl {
     return Status::OK();
   }
 
+  Status Visit(const SparseUnionType& type) { return MakeUnionScalar(type); }
+
+  Status Visit(const DenseUnionType& type) { return MakeUnionScalar(type); }
+
+  template <typename T, typename ScalarType = typename TypeTraits<T>::ScalarType>
+  Status MakeUnionScalar(const T& type) {
+    if (type.num_fields() == 0) {
+      return Status::Invalid("Cannot make scalar of empty union type");
+    }
+    out_ = std::make_shared<ScalarType>(type.type_codes()[0], type_);
```

Comment:
This will be a scalar with is_valid = false, right? Shouldn't this instead be a valid scalar containing a null scalar of one of the children?

Reply:
I see there's the XXX comment below...

Reply:
It was already the case, this is basically adding the type code parameter.
I agree this behaviour is debatable.